instchown.c

```diff
@@ -10,36 +13,66 @@ extern void init_uidgid();
 
 #define FATAL "instchown: fatal: "
 
+static void die_nomem()
+{
+  strerr_die2sys(111,FATAL,"out of memory");
+}
+
+static void ddhome(stralloc *dd, const char *home)
+{
+  const char *denv = env_get("DESTDIR");
+  if (denv)
+    if (!stralloc_copys(dd,denv)) die_nomem();
+
+  if (!stralloc_catb(dd,home,strlen(home))) die_nomem();
+  if (!stralloc_0(dd)) die_nomem();
+}
+
```

Comment:
This appeals intuitively to me, bringing `instchown.c` more in line with `instfiles.c`. It's been a while since I last thought about the use cases here, so I just want to make sure I understand what we had overlooked that is being addressed here.

In my pkgsrc package (of 1.08), `instpackage` is called two times: first [post-build, with `DESTDIR`](https://github.com/schmonz/pkgsrc-qmail/blob/3dbaf5307a58c8ed017124907e4b1c51b12441a9/Makefile#L223), just before binary package creation, and then [post-package-install, without `DESTDIR`](https://github.com/schmonz/pkgsrc-qmail/blob/3dbaf5307a58c8ed017124907e4b1c51b12441a9/INSTALL#L102). (I can't remember why the second `instpackage` invocation hides warnings.)

`instchown` is called only once: [after the binary package has run `instpackage`](https://github.com/schmonz/pkgsrc-qmail/blob/3dbaf5307a58c8ed017124907e4b1c51b12441a9/INSTALL#L103). I can't remember why it wasn't worth invoking sooner -- maybe because [I don't want the queue installed in the staging area](https://github.com/notqmail/notqmail/pull/61)?

I guess my memory is still too fuzzy to finish piecing this together. What's the use case for `instchown` needing to set ownerships and modes on files in the staging area? It's probably very simple. It might even apply to me, once I understand what it is. :-)

Reply:
`instpackage` installes the files, but with default permissions and ownership. `instchown` only works on existing files.

For most packaging systems one does not need `instchown` as you install into a `DESTDIR` and then tell your spec file or however that is called with your packager which permissions/owners to set. When you do a local installation you would run them without `DESTDIR` and everything is fine.

Gentoo however installs into `DESTDIR`, but with the correct permissions, and then moves things over. Traditionally Gentoo has always used the `queuefix` script to correct the permission in `DESTDIR`, this is why noone noticed that this mode is missing.